README
== Description
  The win32-service library allows you to control or create MS Windows services.

== Installation
  gem install win32-service

== Synopsis
  require 'win32/service'

  # Iterate over the available services
  Win32::Service.services do |service|
    p service
  end

== More Documentation
  Please see the documentation in the 'doc' directory, or the gem documentation
  that was installed when you installed this library as a gem.

== Known Issues
=== Problem:
  Service.delete causes "Unable to delete: The specified service has been
  marked for deletion."
 
=== Troubleshooting:
  This can be caused by one of two things. Either you attempted to delete a
  running service without stopping it first, or you have the Services
  Administrative Tool (GUI) open. In the former case, the solution is to first
  stop the service if it's running. In the latter, close the Services GUI
  admin tool before deleting.
	
=== Problem:
  Service.start causes, "The service did not respond to the start or control
  request in a timely fashion."

=== Troubleshooting:
  The best way to debug your services is to wrap your entire Daemon subclass
  in a begin/end block and send error messages to a file. That should give a
  good clue as to the nature of the problem. The most probable culprits are:

  * You forgot to require 'win32/daemon' in your Daemon code.

  * You've tried to require a library that's not in your $LOAD_PATH. Make sure
  that your require statements are inside the begin/rescue block so that you can
  easily find those mistakes.

  * Your have a bad binary path name. Be sure to use an absolute path name for
  the binary path name, including the full path to the Ruby interpreter, e.g.
  'c:\ruby\bin\ruby' instead of just 'ruby'.

  * You've got a syntax error in your code somewhere.

== See Also
  ruby-wmi

== Future Plans
  Add service_session_change hook

== Copyright
  (C) 2003-2016, Daniel J. Berger, All Rights Reserved
    
== License
  Artistic 2.0

== Warranty
  This package is provided "as is" and without any express or
  implied warranties, including, without limitation, the implied
  warranties of merchantability and fitness for a particular purpose.
    
== Authors
  Daniel J. Berger   
  Park Heesob


CHANGES
== 0.8.10 - 11-Dec-2016
* Fixed a bug where action resolution information was not correct. Thanks go
  to BXII-Dave for the spot and patch.

== 0.8.9 - 26-Jun-2016
* Added the ffi-win32-extensions dependency.
* Changed the way service dependencies are determined.
* Removed the helper.rb file since it's no longer used.

== 0.8.8 - 24-Jun-2016
* Fixed the function prototypes for CreateEvent and AdjustTokenPrivileges
  where I was using :bool instead of :int.
* Updated the various FFI module wrapper names so that they're more distinctly
  namespaced. Avoids a clash with the win32-eventlog gem, and possibly others.
* Discovered more places where QueryServiceConfig2 could fail while attempting
  to get failure actions and delayed auto start info. Instead of failing,
  these now set their respective data to nil.
* A timeout value in our main observer thread was increased to potentially
  improve performance. Feedback welcome.
* This gem is now signed.
* Added the win32-service.rb and win32-daemon.rb files for convenience.
* Use require_relative internally.
* Added win32-security as a development dependency. It was already being used
  for tests but now it's actually part of the gemspec.
* Many tests are now skipped unless run with administrative privileges, either
  because they require it, or because they otherwise take too long.
* Gem related tasks in the Rakefile now assume Rubygems 2.x.
* Added an appveyor.yml file.
* The DemoDaemon example will create C:/Tmp if it doesn't already exist.

== 0.8.7 - 14-Jul-2014
* Fixed a bug in the SERVICE_DELAYED_AUTO_START_INFO struct. It now returns 0
  or 1, though it still accepts true or false as arguments. Thanks go to
  Kartik Null Cating-Subramanian for the spot.

== 0.8.6 - 21-Aug-2014
* Some internal changes that make stopping a service more robust. Thanks go
  to Ethan J. Brown for the patches.
* Renamed the constants module to avoid name clashes. Thanks go to Michael
  Smith for the spot.

== 0.8.5 - 2-Jul-2014
* The service type for Service.new no longer defaults to being an interactive
  process. This could cause process credential issues and wasn't very
  useful in practice anyway. Thanks go to Pierre Ynard for the report.
* More tests skipped unless run with admin privileges.
* Use relative_require instead of doing things manually.
* Minor memory efficieny improvements.
* Skip errors on ERROR_RESOURCE_TYPE_NOT_FOUND when getting service config info.
  Thanks go to Rob Reynolds for the patch.

== 0.8.4 - 23-Jan-2014
* Added support for the delayed start startup type. Thanks go to Jim Arnold
  for the patch.
* Fixed the service test that ensures that trying to start an already running
  service will raise an error. Thanks again to Jim Arnold for the patch.

== 0.8.3 - 1-Nov-2013
* Fixed a bug where a service's state could be changed when merely
  interrogating the service. Thanks go to Glenn Sarti / Puppet Labs
  for the spot.
* Fixed a bug where attempting to get the description of a service that
  has no description could raise an error. Now an empty string is returned.

== 0.8.2 - 13-Aug-2013
* Fixed a bug in the Service.start method where it was not handling arguments
  properly. Thanks go to Sean Karlage for the spot.

== 0.8.1 - 2-Aug-2013
* Fixed the failure actions code for the Service#configure method
  so that it is also using FFI.
* An internal fix for the Daemon class was added. In short, you no longer
  need to, nor should you, call exit! in a service_stop method. Thanks go
  to Adrian Candaten for pointing out the issue in general.
* Made FFI functions private.
* Added Rake as a development dependency.
* Removed an internal helper function that was no longer being used.

== 0.8.0 - 19-Jun-2013
* Converted all code to use FFI. This includes the Daemon code, which means
  that it also now works with JRuby.

== 0.7.2 - 7-Sep-2011
* Now works with mingw compiler by skipping __try and __finally if using
  mingw, or using them with mingw if the seh.h header is found.
* The binary gem now contains binaries for both Ruby 1.8.7 and 1.9.2 using
  a simple wrapper.
* Updated some Rake tasks.

== 0.7.1 - 9-Feb-2010
* Compatibility fixes for 1.9.x.
* Fixed a bug where the controls_accepted struct member was always nil. This
  has been fixed and now always returns an array.
* Added 'pre-shutdown' as possible values for the list of controls accepted.
* Changed test-unit from a runtime dependency to a development dependency.
* Some tests were refactored to be more reliable.
* Refactored the Rakefile and gemspec.

== 0.7.0 - 30-Jun-2009
* INTERFACE CHANGE: The Service.new and Service.configure methods now accept
  a single hash argument only. The :service_name key is mandatory, while the
  :host key is optional.
* Fixed a bug in the Service.services method where a segfault would occur if
  the binary_path_name was greater than 260 characters. The buffer size has
  been increased to 1024 characters. Thanks go to Dan Daniels for the spot.
* Altered the Daemon class so that the Ruby interpreter was not invoked within
  the Service_Main function. The functions that close stdin, stdout and stderr
  are now called within the mainloop method. Thanks go to Tim Hanson for the
  spot and the patch.
* Eliminated a minor build warning in daemon.c
* License changed to Artistic 2.0, with corresponding gemspec update.
* Added an uninstall rake task.
* Updated the service_stop method in the demo_daemon.rb file, and added some
  additional comments.
* The test files were renamed.
* Several tests were updated and refactored to use Test::Unit 2.x. That
  library is now a dependency.
* The 'uninstall' Rakefile task is now more verbose.

== 0.6.1 - 1-Jan-2008
* The Service.services method now handles the scenario where a particular
  service may still exist, but its underlying registry entry has been deleted.
  In this case, most ServiceInfo struct members are set to nil and a warning
  is issued.
* RDoc updates.

== 0.6.0 - 25-Nov-2007

ext/win32/daemon.c
#include "ruby.h"
#include <windows.h>
#include <string.h>
#include <stdlib.h>
#include <malloc.h>

#ifdef HAVE_TCHAR_H
#include <tchar.h>
#endif

#ifdef HAVE_SEH_H
#include <seh.h>
#endif

#define WIN32_SERVICE_VERSION "0.7.2"

// Ruby 1.9.x
#ifndef RSTRING_PTR
#define RSTRING_PTR(s) (RSTRING(s)->ptr)
#endif
#ifndef RSTRING_LEN
#define RSTRING_LEN(s) (RSTRING(s)->len)
#endif

#ifndef RARRAY_PTR
#define RARRAY_PTR(a) (RARRAY(a)->ptr)
#endif
#ifndef RARRAY_LEN
#define RARRAY_LEN(a) (RARRAY(a)->len)
#endif

static VALUE cDaemonError;

static HANDLE hThread;
static HANDLE hStartEvent;
static HANDLE hStopEvent;
static HANDLE hStopCompletedEvent;
static SERVICE_STATUS_HANDLE ssh;
static DWORD dwServiceState;
static int Argc;
static VALUE* Argv;

CRITICAL_SECTION csControlCode;
// I happen to know from looking in the header file
// that 0 is not a valid service control code
// so we will use it, the value does not matter
// as long as it will never show up in ServiceCtrl
// - Patrick Hurley
#define IDLE_CONTROL_CODE 0
static int waiting_control_code = IDLE_CONTROL_CODE;

void  WINAPI Service_Main(DWORD dwArgc, LPTSTR *lpszArgv);
void  WINAPI Service_Ctrl(DWORD dwCtrlCode);
void  SetTheServiceStatus(DWORD dwCurrentState,DWORD dwWin32ExitCode,
                          DWORD dwCheckPoint,  DWORD dwWaitHint);

// Return an error code as a string
LPTSTR ErrorDescription(DWORD p_dwError)
{
  HLOCAL hLocal = NULL;
  #ifdef HAVE_TCHAR_H
  static TCHAR ErrStr[1024];
  #else
  static char ErrStr[1024];
  #endif
  int len;

  if (!(len=FormatMessage(
    FORMAT_MESSAGE_ALLOCATE_BUFFER |
    FORMAT_MESSAGE_FROM_SYSTEM |
    FORMAT_MESSAGE_IGNORE_INSERTS,
    NULL,
    p_dwError,
    MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), // Default language
    (LPTSTR)&hLocal,
    0,
    NULL)))
  {
    rb_raise(rb_eStandardError, "unable to format error message");
  }
  memset(ErrStr, 0, sizeof(ErrStr));
  strncpy(ErrStr, (LPTSTR)hLocal, len-2); // remove \r\n
  LocalFree(hLocal);
  return ErrStr;
}

// Called by the service control manager after the call to
// StartServiceCtrlDispatcher.
void WINAPI Service_Main(DWORD dwArgc, LPTSTR *lpszArgv)
{
  // Obtain the name of the service.
  LPTSTR lpszServiceName = lpszArgv[0];

  // Args passed to Service.start
  if(dwArgc > 1){
    unsigned int i;
    Argc = dwArgc - 1;
    Argv = malloc(sizeof(VALUE)*Argc);

    for(i=1; i < dwArgc; i++)
      Argv[i-1] = rb_str_new2(lpszArgv[i]);
  }

  // Register the service ctrl handler.
  ssh = RegisterServiceCtrlHandler(
    lpszServiceName,
    (LPHANDLER_FUNCTION)Service_Ctrl
  );

  // no service to stop, no service handle to notify, nothing to do but exit
  if(ssh == (SERVICE_STATUS_HANDLE)0)
    return;

  // The service has started.
  SetTheServiceStatus(SERVICE_RUNNING, NO_ERROR, 0, 0);

  SetEvent(hStartEvent);

  // Main loop for the service.
  while(WaitForSingleObject(hStopEvent, 1000) != WAIT_OBJECT_0)
  {
  }

  // Main loop for the service.
  while(WaitForSingleObject(hStopCompletedEvent, 1000) != WAIT_OBJECT_0)
  {
  }

  // Stop the service.
  SetTheServiceStatus(SERVICE_STOPPED, NO_ERROR, 0, 0);
}

VALUE Service_Event_Dispatch(VALUE val)
{
  VALUE func,self;
  VALUE result = Qnil;

  if(val!=Qnil) {
    self = RARRAY_PTR(val)[0];
    func = NUM2INT(RARRAY_PTR(val)[1]);

    result = rb_funcall(self,func,0);
  }

  return result;
}

VALUE Ruby_Service_Ctrl(VALUE self){
  while(WaitForSingleObject(hStopEvent,0) == WAIT_TIMEOUT){
#if !defined(__GNUC__) || defined(HAVE_SEH_H)
    __try{
#endif
      EnterCriticalSection(&csControlCode);

      // Check to see if anything interesting has been signaled
      if(waiting_control_code != IDLE_CONTROL_CODE){
        if(waiting_control_code != SERVICE_CONTROL_STOP){
          // If there is a code, create a ruby thread to deal with it
          // this might be over engineering the solution, but I don't
          // want to block Service_Ctrl longer than necessary and the
          // critical section will block it.
          VALUE EventHookHash = rb_ivar_get(self, rb_intern("@event_hooks"));

          if(EventHookHash != Qnil){
            VALUE val = rb_hash_aref(
              EventHookHash,
              INT2NUM(waiting_control_code)
            );

            if(val != Qnil)
              rb_thread_create(Service_Event_Dispatch, (void*) val);
          }
        }
        else{
          break;
        }

        waiting_control_code = IDLE_CONTROL_CODE;
      }
#if !defined(__GNUC__) || defined(HAVE_SEH_H)
    }
    __finally {
#endif
      LeaveCriticalSection(&csControlCode);
#if !defined(__GNUC__) || defined(HAVE_SEH_H)
    }
#endif
    // This is an ugly polling loop, be as polite as possible
    rb_thread_polling();
  }

  // Force service_stop call
  {
    VALUE EventHookHash = rb_ivar_get(self, rb_intern("@event_hooks"));

    if(EventHookHash != Qnil){
      VALUE val = rb_hash_aref(EventHookHash, INT2NUM(SERVICE_CONTROL_STOP));

      if(val!=Qnil)
        rb_thread_create(Service_Event_Dispatch, (void*) val);
    }
  }

  return Qnil;
}

// Handles control signals from the service control manager.
void WINAPI Service_Ctrl(DWORD dwCtrlCode)
{
  DWORD dwState = SERVICE_RUNNING;

#if !defined(__GNUC__) || defined(HAVE_SEH_H)
  __try{
#endif
    EnterCriticalSection(&csControlCode);
    waiting_control_code = dwCtrlCode;
#if !defined(__GNUC__) || defined(HAVE_SEH_H)
  }
  __finally{
#endif
    LeaveCriticalSection(&csControlCode);
#if !defined(__GNUC__) || defined(HAVE_SEH_H)
  }
#endif

  switch(dwCtrlCode)
  {
    case SERVICE_CONTROL_STOP:
      dwState = SERVICE_STOP_PENDING;
      break;
    case SERVICE_CONTROL_SHUTDOWN:
      dwState = SERVICE_STOP_PENDING;
      break;
    case SERVICE_CONTROL_PAUSE:
      dwState = SERVICE_PAUSED;
      break;
    case SERVICE_CONTROL_CONTINUE:
      dwState = SERVICE_RUNNING;
      break;
    case SERVICE_CONTROL_INTERROGATE:
      break;
    default:
      break;
  }

  // Set the status of the service.
  SetTheServiceStatus(dwState, NO_ERROR, 0, 0);

  // Tell service_main thread to stop.
  if ((dwCtrlCode == SERVICE_CONTROL_STOP) ||
     (dwCtrlCode == SERVICE_CONTROL_SHUTDOWN))
  {
    if(!SetEvent(hStopEvent))
      SetTheServiceStatus(SERVICE_STOPPED, GetLastError(), 0, 0);
  }
}

//  Wraps SetServiceStatus.
void SetTheServiceStatus(DWORD dwCurrentState, DWORD dwWin32ExitCode,
                         DWORD dwCheckPoint,   DWORD dwWaitHint)
{
  SERVICE_STATUS ss;  // Current status of the service.

  // Disable control requests until the service is started.
  if(dwCurrentState == SERVICE_START_PENDING){
    ss.dwControlsAccepted = 0;
  }
  else{
    ss.dwControlsAccepted =
      SERVICE_ACCEPT_STOP|SERVICE_ACCEPT_SHUTDOWN|
      SERVICE_ACCEPT_PAUSE_CONTINUE|SERVICE_ACCEPT_SHUTDOWN;
  }

  // Initialize ss structure.
  ss.dwServiceType             = SERVICE_WIN32_OWN_PROCESS;
  ss.dwServiceSpecificExitCode = 0;
  ss.dwCurrentState            = dwCurrentState;
  ss.dwWin32ExitCode           = dwWin32ExitCode;
  ss.dwCheckPoint              = dwCheckPoint;
  ss.dwWaitHint                = dwWaitHint;

  dwServiceState = dwCurrentState;

  // Send status of the service to the Service Controller.
  if(!SetServiceStatus(ssh, &ss))
    SetEvent(hStopEvent);
}

DWORD WINAPI ThreadProc(LPVOID lpParameter){
  SERVICE_TABLE_ENTRY ste[] =
    {{TEXT(""),(LPSERVICE_MAIN_FUNCTION)Service_Main}, {NULL, NULL}};

  // No service to step, no service handle, no ruby exceptions, just
  // terminate the thread.
  if(!StartServiceCtrlDispatcher(ste))
    return 1;

  return 0;
}

static VALUE daemon_allocate(VALUE klass){
  return Data_Wrap_Struct(klass, 0, 0, 0);
}

// Call service_main method
static VALUE daemon_mainloop_protect(VALUE self)
{
  if(rb_respond_to(self,rb_intern("service_main"))){
    if(Argc == 0)
      rb_funcall(self, rb_intern("service_main"), 0);
    else
      rb_funcall2(self, rb_intern("service_main"), Argc, Argv);
  }

  return self;
}

static VALUE daemon_mainloop_ensure(VALUE self)
{
  int i;

  // Signal both the ruby thread and service_main thread to terminate
  SetEvent(hStopEvent);

  // Wait for ALL ruby threads to exit
  for(i=1; TRUE; i++)
  {
    VALUE list = rb_funcall(rb_cThread, rb_intern("list"), 0);

    if(RARRAY_LEN(list) <= 1)
      break;

    // This is another ugly polling loop, be as polite as possible
    rb_thread_polling();

    SetTheServiceStatus(SERVICE_STOP_PENDING, 0, i, 1000);
  }

  // Only one ruby thread
  SetEvent(hStopCompletedEvent);

  // Wait for the thread to stop BEFORE we close the hStopEvent handle
  WaitForSingleObject(hThread, INFINITE);

  // Close the event handle, ignoring failures. We may be cleaning up
  // after an exception, so let that exception fall through.
  CloseHandle(hStopEvent);

  return self;
}

/*
 * This is the method that actually puts your code into a loop and allows it
 * to run as a service.  The code that is actually run while in the mainloop
 * is what you defined in your own Daemon#service_main method.
 */
static VALUE daemon_mainloop(VALUE self)
{
  DWORD ThreadId;
  HANDLE events[2];
  DWORD index;
  VALUE result, EventHookHash;
  int status = 0;

  dwServiceState = 0;

  // Redirect STDIN, STDOUT and STDERR to the NUL device if they're still
  // associated with a tty. This helps newbs avoid Errno::EBADF errors.
  if(rb_funcall(rb_stdin, rb_intern("isatty"), 0) == Qtrue)
    rb_funcall(rb_stdin, rb_intern("reopen"), 1, rb_str_new2("NUL"));

  if(rb_funcall(rb_stdout, rb_intern("isatty"), 0) == Qtrue)
    rb_funcall(rb_stdout, rb_intern("reopen"), 1, rb_str_new2("NUL"));

  if(rb_funcall(rb_stderr, rb_intern("isatty"), 0) == Qtrue)
    rb_funcall(rb_stderr, rb_intern("reopen"), 1, rb_str_new2("NUL"));

  // Use a markable instance variable to prevent the garbage collector
  // from freeing the hash before Ruby_Service_Ctrl exits, or just
  // at any ole time while running the service
  EventHookHash = rb_hash_new();
  rb_ivar_set(self, rb_intern("@event_hooks"), EventHookHash);

  // Event hooks
  if(rb_respond_to(self, rb_intern("service_stop"))){
    rb_hash_aset(EventHookHash, INT2NUM(SERVICE_CONTROL_STOP),
      rb_ary_new3(2, self, INT2NUM(rb_intern("service_stop"))));
  }

  if(rb_respond_to(self, rb_intern("service_pause"))){
    rb_hash_aset(EventHookHash, INT2NUM(SERVICE_CONTROL_PAUSE),
      rb_ary_new3(2, self, INT2NUM(rb_intern("service_pause"))));
  }

  if(rb_respond_to(self, rb_intern("service_resume"))){
    rb_hash_aset(EventHookHash, INT2NUM(SERVICE_CONTROL_CONTINUE),
      rb_ary_new3(2, self, INT2NUM(rb_intern("service_resume"))));
  }

  if(rb_respond_to(self, rb_intern("service_interrogate"))){
    rb_hash_aset(EventHookHash, INT2NUM(SERVICE_CONTROL_INTERROGATE),
      rb_ary_new3(2, self, INT2NUM(rb_intern("service_interrogate"))));
  }

  if(rb_respond_to(self, rb_intern("service_shutdown"))){
    rb_hash_aset(EventHookHash, INT2NUM(SERVICE_CONTROL_SHUTDOWN),
      rb_ary_new3(2, self, INT2NUM(rb_intern("service_shutdown"))));
  }

#ifdef SERVICE_CONTROL_PARAMCHANGE
  if(rb_respond_to(self, rb_intern("service_paramchange"))){
    rb_hash_aset(EventHookHash, INT2NUM(SERVICE_CONTROL_PARAMCHANGE),
      rb_ary_new3(2, self, INT2NUM(rb_intern("service_paramchange"))));
  }
#endif

#ifdef SERVICE_CONTROL_NETBINDADD
  if(rb_respond_to(self, rb_intern("service_netbindadd"))){
    rb_hash_aset(EventHookHash, INT2NUM(SERVICE_CONTROL_NETBINDADD),
      rb_ary_new3(2, self, INT2NUM(rb_intern("service_netbindadd"))));
  }
#endif

#ifdef SERVICE_CONTROL_NETBINDREMOVE
  if(rb_respond_to(self, rb_intern("service_netbindremove"))){
    rb_hash_aset(EventHookHash, INT2NUM(SERVICE_CONTROL_NETBINDREMOVE),
      rb_ary_new3(2, self, INT2NUM(rb_intern("service_netbindremove"))));
  }
#endif

#ifdef SERVICE_CONTROL_NETBINDENABLE
  if(rb_respond_to(self, rb_intern("service_netbindenable"))){
    rb_hash_aset(EventHookHash, INT2NUM(SERVICE_CONTROL_NETBINDENABLE),
      rb_ary_new3(2, self, INT2NUM(rb_intern("service_netbindenable"))));
  }
#endif

#ifdef SERVICE_CONTROL_NETBINDDISABLE
  if(rb_respond_to(self, rb_intern("service_netbinddisable"))){
    rb_hash_aset(EventHookHash, INT2NUM(SERVICE_CONTROL_NETBINDDISABLE),
      rb_ary_new3(2, self, INT2NUM(rb_intern("service_netbinddisable"))));
  }
#endif

  // Calling init here so that init failures never even tries to
  // start the service... of course that means that init methods
  // must be very quick, because the SCM will be receiving no
  // START_PENDING messages while init's running - I may fix this
  // later
  if(rb_respond_to(self, rb_intern("service_init")))
    rb_funcall(self, rb_intern("service_init"),0);

  // Create the event to signal the service to start.
  hStartEvent = CreateEvent(NULL, TRUE, FALSE, NULL);

  if(hStartEvent == NULL)
    rb_raise(cDaemonError, ErrorDescription(GetLastError()));

  // Create the event to signal the service to stop.
  hStopEvent = CreateEvent(NULL, TRUE, FALSE, NULL);

  if(hStopEvent == NULL)
    rb_raise(cDaemonError, ErrorDescription(GetLastError()));

  // Create the event to signal the service that stop has completed
  hStopCompletedEvent = CreateEvent(NULL, TRUE, FALSE, NULL);

  if(hStopCompletedEvent == NULL)
    rb_raise(cDaemonError, ErrorDescription(GetLastError()));

  // Create Thread for service main
  hThread = CreateThread(NULL, 0, ThreadProc, 0, 0, &ThreadId);

  if(hThread == INVALID_HANDLE_VALUE)
    rb_raise(cDaemonError, ErrorDescription(GetLastError()));

  events[0] = hThread;
  events[1] = hStartEvent;

  // wait for Service_Main function to either start the service OR terminate
  while((index = WaitForMultipleObjects(2,events,FALSE,1000)) == WAIT_TIMEOUT)
  {
  }

  // thread exited, so the show is off
  if(index == WAIT_OBJECT_0)
    rb_raise(cDaemonError, "Service_Main thread exited abnormally");

  // from this point onward, stopevent must be triggered!

  // Create the green thread to poll for Service_Ctrl events
  rb_thread_create(Ruby_Service_Ctrl, (void *)self);

  result = rb_protect(daemon_mainloop_protect, self, &status);

  // service_main raised an exception
  if(status){
    daemon_mainloop_ensure(self);
    rb_jump_tag(status);
  }

  // service_main exited cleanly
  return daemon_mainloop_ensure(self);
}

/*
 * Returns the state of the service (as an constant integer) which can be any
 * of the service status constants, e.g. RUNNING, PAUSED, etc.
 *
 * This method is typically used within your service_main method to setup the
 * loop. For example:
 *
 *    class MyDaemon < Daemon
 *       def service_main
 *          while state == RUNNING || state == PAUSED || state == IDLE
 *             # Your main loop here
 *          end
 *       end
 *    end
 *
 * See the Daemon#running? method for an abstraction of the above code.
 */
static VALUE daemon_state(VALUE self){
  return UINT2NUM(dwServiceState);
}

/*
 * Returns whether or not the service is in a running state, i.e. the service
 * status is either RUNNING, PAUSED or IDLE.
 *
 * This is typically used within your service_main method to setup the main
 * loop. For example:
 *
 *    class MyDaemon < Daemon
 *       def service_main
 *          while running?
 *             # Your main loop here
 *          end
 *       end
 *    end
 */
static VALUE daemon_is_running(VALUE self){
  VALUE v_bool = Qfalse;

  if(
    (dwServiceState == SERVICE_RUNNING) ||
    (dwServiceState == SERVICE_PAUSED) ||
    (dwServiceState == 0)
  ){
    v_bool = Qtrue;
  }

  return v_bool;
}

/*
 * This is a shortcut for Daemon.new + Daemon#mainloop.
 */
static VALUE daemon_c_mainloop(VALUE klass){
  VALUE v_args[1];
  VALUE v_daemon = rb_class_new_instance(0, v_args, klass);
  return rb_funcall(v_daemon, rb_intern("mainloop"), 0, 0);
}

void Init_daemon()
{
  /* The Win32 module serves as a namespace only. */
  VALUE mWin32 = rb_define_module("Win32");

  /* The Daemon class encapsulates a Windows service through the use
   * of callback methods and a main loop.
   */
  VALUE cDaemon = rb_define_class_under(mWin32, "Daemon", rb_cObject);

  /* Error typically raised if something goes wrong with your daemon. */
  cDaemonError = rb_define_class_under(cDaemon, "Error", rb_eStandardError);

  rb_define_alloc_func(cDaemon, daemon_allocate);
  rb_define_method(cDaemon, "mainloop", daemon_mainloop, 0);
  rb_define_method(cDaemon, "state", daemon_state, 0);
  rb_define_method(cDaemon, "running?", daemon_is_running, 0);

  rb_define_singleton_method(cDaemon, "mainloop", daemon_c_mainloop, 0);

  // Intialize critical section used by green polling thread
  InitializeCriticalSection(&csControlCode);

  // Constants

  /* 0.7.3: The version of this library */
  rb_define_const(cDaemon, "VERSION", rb_str_new2(WIN32_SERVICE_VERSION));

  /* Service has received a signal to resume but is not yet running */
  rb_define_const(cDaemon, "CONTINUE_PENDING",
    INT2NUM(SERVICE_CONTINUE_PENDING));

  /* Service has received a signal to pause but is not yet paused */
  rb_define_const(cDaemon, "PAUSE_PENDING", INT2NUM(SERVICE_PAUSE_PENDING));

  /* Service is in a paused state */
  rb_define_const(cDaemon, "PAUSED", INT2NUM(SERVICE_PAUSED));

  /* Service is running */
  rb_define_const(cDaemon, "RUNNING", INT2NUM(SERVICE_RUNNING));

  /* Service has received a signal to start but is not yet running */
  rb_define_const(cDaemon, "START_PENDING", INT2NUM(SERVICE_START_PENDING));

  /* Service has received a signal to stop but has not yet stopped */
  rb_define_const(cDaemon, "STOP_PENDING", INT2NUM(SERVICE_STOP_PENDING));

  /* Service is stopped. */
  rb_define_const(cDaemon, "STOPPED", INT2NUM(SERVICE_STOPPED));

  /* Service is in an idle state */
  rb_define_const(cDaemon, "IDLE", INT2NUM(0));
}
